Userland/Services/RequestServer/ConnectionFromClient.cpp

```diff
@@ -26,8 +26,140 @@ static IDAllocator s_client_ids;
 
 ConnectionFromClient::ConnectionFromClient(NonnullOwnPtr<Core::LocalSocket> socket)
     : IPC::ConnectionFromClient<RequestClientEndpoint, RequestServerEndpoint>(*this, move(socket), s_client_ids.allocate())
+    , m_work_queue(MUST((Core::SharedSingleProducerCircularQueue<Work, 256>::create())))
 {
     s_connections.set(client_id(), *this);
+    for (size_t i = 0; i < Core::System::hardware_concurrency(); ++i)
+        m_connection_workers.append(MUST(Threading::WorkerThread<Error>::create("RequestServer Worker"sv)));
+
+    for (auto& connection_worker : m_connection_workers) {
+        connection_worker->start_task([this] {
+            return worker_main();
+        });
+    }
+}
+
+class Job : public RefCounted<Job>
+    , public Weakable<Job> {
+public:
+    static NonnullRefPtr<Job> ensure(URL::URL const& url)
+    {
+        RefPtr<Job> job;
+        if (auto it = s_jobs.find(url); it != s_jobs.end())
+            job = it->value.strong_ref();
+        if (job == nullptr) {
+            job = adopt_ref(*new Job(url));
+            s_jobs.set(url, job);
+        }
+        return *job;
+    }
+
+    void start(Core::Socket& socket)
+    {
+        auto is_connected = socket.is_open();
+        VERIFY(is_connected);
+        ConnectionCache::request_did_finish(m_url, &socket);
+    }
+
+    void fail(Core::NetworkJob::Error error)
+    {
+        dbgln("Pre-connect to {} failed: {}", m_url, Core::to_string(error));
+    }
+
+    void will_be_destroyed() const
+    {
+        s_jobs.remove(m_url);
+    }
+
+private:
+    explicit Job(URL::URL url)
+        : m_url(move(url))
+    {
+    }
+
+    URL::URL m_url;
+    inline static HashMap<URL::URL, WeakPtr<Job>> s_jobs {};
+};
+
+ErrorOr<void> ConnectionFromClient::worker_main()
+{
+    Core::EventLoop thread_loop;
+
+    NonnullRefPtr<Core::Timer> timer = Core::Timer::create_repeating(100, [&] {
+        // Don't restart this function while we're processing.
+        timer->stop();
+        ScopeGuard guard([&] {
+            timer->start();
+        });
+
+        do {
+            if (!m_work_queue.is_valid())
```

Comment:
what's an "invalid" work queue :thinking:

Reply:
deleted on the other side, it's...weird.